src/wallet/wallet.cpp

```diff
@@ -522,6 +522,121 @@ void CWallet::UpgradeDescriptorCache()
+
+    for (const auto& spkm : GetAllScriptPubKeyMans()) {
+        const DescriptorScriptPubKeyMan* desc_spkm = dynamic_cast<DescriptorScriptPubKeyMan*>(spkm);
+        assert(desc_spkm);
+        LOCK(desc_spkm->cs_desc_man);
+        WalletDescriptor w_desc = desc_spkm->GetWalletDescriptor();
+
+        // Automatically generated descriptors are ranged
+        if (!w_desc.descriptor->IsRange()) {
+            continue;
+        }
+
+        // Audotmatically generated descriptors have exactly 1 xpub and no other keys
+        std::set<CPubKey> desc_pubkeys;
+        std::set<CExtPubKey> desc_xpubs;
+        w_desc.descriptor->GetPubKeys(desc_pubkeys, desc_xpubs);
+        if (desc_xpubs.size() != 1 || desc_pubkeys.size() != 0) {
+            continue;
+        }
+        const CExtPubKey& xpub = *desc_xpubs.begin();
+
+        if (descs_keys.count(xpub) == 0) {
+            descs_keys.emplace(xpub, std::make_pair(tmpl, 0));
+        }
+
+        // Automatically generated descriptors have an output type that is one of legacy, p2sh-segwit, or bech32
+        std::optional<OutputType> output_type = w_desc.descriptor->GetOutputType();
+        if (!output_type.has_value()) {
+            continue;
+        }
+        if (LEGACY_OUTPUT_TYPES.count(output_type.value()) == 0) {
+            continue;
+        }
+        descs_keys[xpub].first[output_type.value()]++;
+
+        if (w_desc.creation_time > descs_keys[xpub].second) {
+            descs_keys[xpub].second = w_desc.creation_time;
+        }
+    }
+
+    // Find candidate active xpubs
+    // These are the ones that are used in 2 pkh(), 2 sh(wpkh()), and 2 wpkh() descriptors
+    uint64_t best_time = 0;
+    std::optional<CExtPubKey> best_xpub;
+    std::map<CKeyID, CExtPubKey> xpubs;
+    for (const auto& [xpub, info] : descs_keys) {
+        const auto& [dtypes, desc_time] = info;
+        if (std::all_of(dtypes.begin(), dtypes.end(), [](const auto& p) { return p.second == 2;})) {
+            xpubs.emplace(xpub.pubkey.GetID(), xpub);
+            if (desc_time > best_time) {
+                best_time = desc_time;
+                best_xpub.emplace(xpub);
+            }
+        }
+    }
+
+    // Add hd keys
+    for (const auto& [id_pair, key] : desc_keys) {
+        const auto& [desc_id, keyid] = id_pair;
+
+        const auto& it = xpubs.find(keyid);
+        if (it == xpubs.end()) {
+            continue;
+        }
+
+        CExtKey extkey(it->second, key);
+        AddHDKey(extkey);
+        xpubs.erase(it);
+    }
+    for (const auto& [id_pair, key_pair] : desc_crypt_keys) {
+        const auto& [desc_id, keyid] = id_pair;
+        const auto& [pubkey, ckey] = key_pair;
+
+        const auto& it = xpubs.find(keyid);
+        if (it == xpubs.end()) {
+            continue;
+        }
+
+        LoadHDCryptedKey(it->second, ckey);
+        batch.WriteHDCryptedKey(it->second, ckey);
```

Comment:
Yes this is nice to have. Since the key format could never change due to backward compatibility requirements, this is not an issue. You can resolve this comment.